File: Stack/173_BinarySearchTreeIterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stk = []
        self.pushAll(root)

    def hasNext(self):
        """
        :rtype: bool
        """
        return len(self.stk) > 0

    def next(self):
        """
        :rtype: int
        """
        node = self.stk.pop()
        self.pushAll(node.right)
        return node.val

    def pushAll(self, node):
        while node is not None:
            self.stk.append(node)
            node = node.left

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())

File: Stack/173_BinarySearchTreeIterator.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class BSTIterator {

    private ArrayList<TreeNode> lst;
    private int index;
    public BSTIterator(TreeNode root) {
        lst = new ArrayList<>();
        index = 0;
        Stack<TreeNode> stk = new Stack<>();
        TreeNode node = root;
        while(!stk.isEmpty() || node != null) {
            if (node == null) {
                node = stk.pop();
                lst.add(node);
                node = node.right;
            } else {
                stk.push(node);
                node = node.left;
            }
        }
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        if (index < lst.size()) {
            return true;
        }
        return false;
    }

    /** @return the next smallest number */
    public int next() {
        return lst.get(index++).val;
    }
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = new BSTIterator(root);
 * while (i.hasNext()) v[f()] = i.next();
 */